### PR TITLE
Implement CountRows for fast path of countings

### LINKED ISSUE #219: Implement Lance CountRows to enable fast path to count records.

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #219

### CODE REVIEW

File: cpp/src/lance/arrow/arrow_dataset_test.cc [hunk 1]
```diff
@@ -58,3 +63,18 @@ TEST_CASE("FileSystemFactory Test") {
...
+  auto uri = std::string("file://") + path.string();
+
+  auto schema = arrow::schema({arrow::field("key", arrow::utf8())});
+  std::vector<int32_t> values(1000);
+  std::iota(std::begin(values), std::end(values), 1);
+  auto table = ::arrow::Table::Make(schema, {ToArray(values).ValueOrDie()});
+
+  auto dataset = MakeDataset(table, {}, 32).ValueOrDie();
+  auto scanner = dataset->NewScan().ValueOrDie()->Finish().ValueOrDie();
+  CHECK(scanner->CountRows().ValueOrDie() == 1000);
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/arrow_dataset_test.cc:79]: Anyway to check that the fast core path is hit?

**Review by changhiskhan — COMMENTED** [review]
Small nits/questions

File: cpp/src/lance/arrow/file_lance.cc [hunk 2]
```diff
@@ -71,6 +71,31 @@ ::arrow::Result<std::shared_ptr<::arrow::Schema>> LanceFileFormat::Inspect(
...
+  if (predicate.Equals(::arrow::compute::literal(true))) {
+    // Fast path.
+    auto executor = options->io_context.executor();
+    assert(executor != nullptr);
+    auto result = executor->Submit(
+        [&](const auto& file) -> ::arrow::Result<::arrow::util::optional<int64_t>> {
+          ARROW_ASSIGN_OR_RAISE(auto infile, file->source().Open());
+          ARROW_ASSIGN_OR_RAISE(auto reader,
+                                lance::io::FileReader::Make(infile, this->impl_->manifest));
+          return ::arrow::util::make_optional(reader->length());
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/arrow/file_lance.cc:87]: This is returning the number of "rows" in the dataset. It does not read the actual data.

File: cpp/src/lance/arrow/file_lance.cc [hunk 3]
```diff
@@ -71,6 +71,31 @@ ::arrow::Result<std::shared_ptr<::arrow::Schema>> LanceFileFormat::Inspect(
   return impl_->manifest->schema().ToArrow();
 }
 
+::arrow::Future<::arrow::util::optional<int64_t>> LanceFileFormat::CountRows(
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply cpp/src/lance/arrow/file_lance.cc:74]: `pyarrow` calls C++ Scanner.CountRows() which eventually calls this method.

File: cpp/src/lance/arrow/arrow_dataset_test.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/arrow_dataset_test.cc:79]: I've added some print messages to make sure path is hit.

To have appropriate tests, i guess we can either make some dependency ingestions into the class, or, doing some surgery that writes a file with "correct metadata", but no data? 

Or, set up some code coverage tool in CI?

Should we do that in a follow up ticket?

Comment (changhiskhan) [reply cpp/src/lance/arrow/arrow_dataset_test.cc:79]: Yeah I'm ok with that, I guess for now just verify by hand if you do count rows on coco it doesn't take forever :)

**Review by changhiskhan — APPROVED** [review]

**Pull request merged**